### PR TITLE
Async full-refresh for EInkDynamicDisplay

### PR DESCRIPTION
**Pull request by todd-herbert**
### Aim
Minimize time where display code blocks execution, by running full-refresh asynchronously.

### Code summary
[A slight modification to GxEPD2](https://github.com/todd-herbert/meshtastic-GxEPD2/commit/8234d199859537768c5cd106bdc3c4901b1ece65) splits the `nextPage()` method into two parts (if variant uses EInkDynamicDisplay). The EInkDisplay base-class calls `nextPage()` as normal. If GxEPD2 detects a full refresh, it bypasses both the `_WaitWhileBusy()` method, and the remaining code of `nextPage()`.

The EInkDynamicDisplay class sets a flag, then returns control to the loop(). As always, the next loop requests a new frame by calling `EInkDynamicDisplay::display()`. While processing new frames, the EInkDynamicDisplay class polls the display's BUSY pin. If it detects a completed async refresh, it calls the new `GxEPD2_BW::endAsyncFull()` method, which contains the post-update code transplanted from the original nextPage() method.

### Prior to merge
* Would like to hear if this does address https://github.com/meshtastic/firmware/issues/2356, mentioned recently by @GUVWAF
* Will create pull-request for the GxEPD2 modifications
* Will adjust `lib_deps` for Wireless Paper to point back at meshtastic/GxEPD2 (currently tracking personal repo, for draft)

### Flow diagram
To aid future maintenance. Traces flow of an async full-refresh between meshtastic/firmware and modified GxEPD2.
flow diagram for an async full refresh

### CODE REVIEW

**Comment by GUVWAF** [comment]
Just tested this on the Wireless Paper v1.1 while sending multiple 1-character text messages from another device in the same room quickly after each other. Previously it would sometimes not get an acknowledgment when I saw it doing a full refresh. With this, I can’t reproduce this anymore, even when it’s doing the full refresh.

Really nice to finally have a complete solution to this problem! The approach looks solid to me. 

Do you know if the e-paper of the T-Echo also has this possibility? Currently we’re using only partial updates on that hardware.

**Comment by todd-herbert** [comment]
> Do you know if the e-paper of the T-Echo also has this possibility? Currently we’re using only fast updates on that hardware.

Hopefully it should be as simple as configuring the macros in platformio.ini for the variant, but I hadn't dared to set it for a bunch of devices I can't test first.

I'll push the same config used for the Wireless Paper v1.1 into T-Echo's platformio.ini; let me know if it works!

The values might want some adjusting to get best performance from the T-Echo's display. 
There's a brief description of the config options in https://github.com/meshtastic/firmware/pull/3316.
If you do optimize the config, let me know and I can push it from here before merge if you'd like. 
Not that it needs saying, but please also feel free to adjust config / implement for any other boards and open independent pull requests with your results!

File: variants/t-echo/platformio.ini [hunk 1]
```diff
@@ -11,6 +11,12 @@ build_flags = ${nrf52840_base.build_flags} -Ivariants/t-echo
   -DEINK_DISPLAY_MODEL=GxEPD2_154_D67
   -DEINK_WIDTH=200
   -DEINK_HEIGHT=200
+  -D USE_EINK_DYNAMICDISPLAY            ; Enable Dynamic EInk
```
Comment (GUVWAF) [inline variants/t-echo/platformio.ini:14]: Maybe remove the space to be consistent with the rest.

**Comment by GUVWAF** [comment]
> I've just pushed the same config used for the Wireless Paper v1.1 into T-Echo's platformio.ini; let me know if it works!

Oh, it’s really that easy -- I really like the modular approach.

Also on the T-Echo I don’t miss any packets even when it does the full update. 

I did not experience a lot of ghosting yet, so maybe we can increase the `EINK_LIMIT_FASTREFRESH` to 15 or 20? The rest seems okay for now.

File: variants/t-echo/platformio.ini (hunk 1 above)
Comment (todd-herbert) [reply variants/t-echo/platformio.ini]: Ha, I did spot this before, and it did bother me; no objections here! (7275c21f6b054e57a9c2c1b8a418cb1aa047bab5)

**Comment by todd-herbert** [comment]
>I did not experience a lot of ghosting yet, so maybe we can increase the EINK_LIMIT_FASTREFRESH to 15 or 20? The rest seems okay for now.

Will do!

> Edit: One thing I noticed is that when shutting down, it stays on the page saying "Shutting down..." instead of going to the screen that says "Sleeping...". Also when booting, it looks like it skips the boot screen. This also happens on the Wireless Paper.

Ah well spotted. I *can* see where the issues is with shutdown, and I think I can see what might be different about the T-Echo's boot right now. I can't reproduce boot screen issues with the Wireless Paper though. Can you confirm that the boot screen is skipped for both the T-Echo and the Wire?

Would it be possible to ask for your serial debug log? It might help be hone-in on the problem.

**Comment by GUVWAF** [comment]
Yes, for me skips the boot screen on both the T-Echo and the Wireless Paper v1.1. It's fully white before showing the node list.

Here is the boot log from the T-Echo (from e-ink init):
> INFO  | 21:26:42 4 Doing EInk init
> DEBUG | 21:26:42 4 determineMode(): ghostPixels=0, refresh=SKIPPED, reason=FRAME_MATCHED_PREVIOUS
> INFO  | 21:26:42 4 Turning on screen
> DEBUG | 21:26:42 4 Module wants a UI Frame
> WARN  | 21:26:42 4 SX126xInterface(cs=24, irq=20, rst=25, busy=17)
> DEBUG | 21:26:42 4 SX126X_DIO3_TCXO_VOLTAGE defined, using DIO3 as TCXO reference voltage at 1.800000 V
> INFO  | 21:26:42 4 Starting meshradio init...
> DEBUG | 21:26:42 4 (bw=250, sf=11, cr=4/5) packet symLen=8 ms, payloadSize=0, time 231 ms
> DEBUG | 21:26:42 4 (bw=250, sf=11, cr=4/5) packet symLen=8 ms, payloadSize=253, time 2115 ms
> INFO  | 21:26:42 4 Radio freq=869.525, config.lora.frequency_offset=0.000
> INFO  | 21:26:42 4 Set radio: region=EU_868, name=LongFast, config=0, ch=0, power=27
> INFO  | 21:26:42 4 Radio myRegion->freqStart -> myRegion->freqEnd: 869.400024 -> 869.650024 (0.250000 mhz)
> INFO  | 21:26:42 4 Radio myRegion->numChannels: 1 x 250.000kHz
> INFO  | 21:26:42 4 Radio channel_num: 1
> INFO  | 21:26:42 4 Radio frequency: 869.525024
> INFO  | 21:26:42 4 Slot time: 42 msec
> INFO  | 21:26:42 4 Set radio: final power level=22
> INFO  | 21:26:42 4 SX126x init result 0
> INFO  | 21:26:42 4 Frequency set to 869.525024
> INFO  | 21:26:42 4 Bandwidth set to 250.000000
> INFO  | 21:26:42 4 Power output set to 22
> DEBUG | 21:26:42 4 Current limit set to 140.000000
> DEBUG | 21:26:42 4 Current limit set result 0
> DEBUG | 21:26:42 4 Setting DIO2 as RF switch
> DEBUG | 21:26:42 4 SX126X_RXEN not defined, defaulting to RADIOLIB_NC
> DEBUG | 21:26:42 4 SX126X_TXEN not defined, defaulting to RADIOLIB_NC
> DEBUG | 21:26:42 4 Using MCU pin -1 as RXEN and pin -1 as TXEN to control RF switching
> INFO  | 21:26:42 4 Set RX gain to boosted mode; result: 0
> INFO  | 21:26:42 4 SX1262 Radio init succeeded, using SX1262 radio
> DEBUG | 21:26:42 4 (bw=250, sf=11, cr=4/5) packet symLen=8 ms, payloadSize=237, time 2000 ms
> DEBUG | 21:26:42 4 LoRA bitrate = 118.500000 bytes / sec
> INFO  | 21:26:42 4 PowerFSM init, USB power=1
> DEBUG | 21:26:42 4 Enter state: BOOT
> DEBUG | 21:26:42 4 [Power] Battery: usbPower=0, isCharging=0, batMv=3100, batPct=0
> DEBUG | 21:26:42 4 [Screen] Screen: Started...
> DEBUG | 21:26:42 4 [GPS] Probing for GPS at 9600 
> INFO  | 21:26:42 4 [GPS] L76K GNSS init succeeded, using L76K GNSS Module
> DEBUG | 21:26:43 5 [GPS] NMEA GPS time 2024-03-08 21:27:49
> DEBUG | 21:26:43 5 [GPS] Upgrading time to quality 4
> DEBUG | 21:27:49 5 [GPS] PCF8563_RTC setDateTime 2024-03-08 21:27:49 536954100
> DEBUG | 21:27:49 5 [GPS] Read RTC time from PCF8563 as 536901740
> DEBUG | 21:27:49 5 [GPS] publishing pos@0:2, hasVal=0, Sats=0, GPSlock=0
> DEBUG | 21:27:49 5 [GPS] No GPS lock
> DEBUG | 21:27:49 5 [GPS] onGPSChanged() pos@0, time=1709933269, lat=0, lon=0, alt=0
> INFO  | 21:27:49 5 [GPS] updatePosition LOCAL pos@0, time=1709933269, latI=0, lonI=0, alt=0
> DEBUG | 21:27:49 5 [GPS] Setting local position: latitude=0, longitude=0, time=1709933269
> DEBUG | 21:27:49 5 [GPS] Node status update: 1 online, 13 total
> INFO  | 21:27:49 5 [RangeTestModule] Range Test Module - Disabled
> DEBUG | 21:27:49 5 [Screen] determineMode(): ghostPixels=0, refresh=FAST, reason=NO_OBJECTIONS, fastRefreshCount=0
> DEBUG | 21:27:49 5 [Screen] Updating E-Paper... done
> INFO  | 21:27:51 8 [PowerFSM] Loss of power in Powered
> INFO  | 21:27:51 8 [PowerFSM] Loss of power in Powered
> DEBUG | 21:27:51 8 [PowerFSM] Initializing NRF52 Bluetooth
> INFO  | 21:27:51 8 [PowerFSM] Initialize the Bluefruit nRF52 module
> INFO  | 21:27:51 8 [PowerFSM] Bluetooth pin set to '333061'
> INFO  | 21:27:51 8 [PowerFSM] Configuring the Device Information Service
> INFO  | 21:27:51 8 [PowerFSM] Configuring the Battery Service
> INFO  | 21:27:51 8 [PowerFSM] Configuring the Mesh bluetooth service
> INFO  | 21:27:51 8 [PowerFSM] Setting up the advertising payload(s)
> INFO  | 21:27:51 8 [PowerFSM] Advertising
> DEBUG | 21:27:51 8 [PowerFSM] Enter state: ON
> DEBUG | 21:27:51 8 [Screen] Screen: Unpowered...
> DEBUG | 21:27:51 8 [Screen] determineMode(): ghostPixels=0, refresh=SKIPPED, reason=FRAME_MATCHED_PREVIOUS
> INFO  | 21:27:52 9 [Screen] Done with boot screen...
> DEBUG | 21:27:52 9 [Screen] showing standard frames
> DEBUG | 21:27:52 9 [Screen] Showing 0 module frames
> DEBUG | 21:27:52 9 [Screen] Total frame count: 103
> DEBUG | 21:27:52 9 [Screen] Added modules.  numframes: 0
> DEBUG | 21:27:52 9 [Screen] Finished building frames. numframes: 6
> DEBUG | 21:27:52 9 [Screen] determineMode(): ghostPixels=1277, refresh=FAST, reason=NO_OBJECTIONS, fastRefreshCount=1
> DEBUG | 21:27:53 9 [Screen] Updating E-Paper... done

And on shutdown: 
> DEBUG | 21:31:34 231 [Screen] showing shutdown screen
> DEBUG | 21:31:34 231 [Screen] determineMode(): ghostPixels=2860, refresh=FULL, reason=EXCEEDED_GHOSTINGLIMIT
> DEBUG | 21:31:35 231 [Screen] Updating E-Paper... done
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] determineMode(): ghostPixels=0, refresh=SKIPPED, reason=FRAME_MATCHED_PREVIOUS
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:37 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 234 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> DEBUG | 21:31:38 235 [Screen] showing shutdown screen
> INFO  | 21:31:38 235 Shutting down from admin command
> INFO  | 21:31:38 235 Turning off screen
> INFO  | 21:31:38 235 Shutting down
> INFO  | 21:31:38 235 Entering deep sleep forever

> And the boot log of the Wireless Paper v1.1:
> DEBUG | ??:??:?? 5 determineMode(): ghostPixels=0, refresh=SKIPPED, reason=FRAME_MATCHED_PREVIOUS
> INFO  | ??:??:?? 5 Turning on screen
> DEBUG | ??:??:?? 5 Module wants a UI Frame
> WARN  | ??:??:?? 5 SX126xInterface(cs=8, irq=14, rst=12, busy=13)
> DEBUG | ??:??:?? 5 SX126X_DIO3_TCXO_VOLTAGE defined, using DIO3 as TCXO reference voltage at 1.800000 V
> INFO  | ??:??:?? 5 Starting meshradio init...
> DEBUG | ??:??:?? 5 (bw=250, sf=11, cr=4/5) packet symLen=8 ms, payloadSize=0, time 231 ms
> DEBUG | ??:??:?? 5 (bw=250, sf=11, cr=4/5) packet symLen=8 ms, payloadSize=253, time 2115 ms
> INFO  | ??:??:?? 5 Radio freq=869.525, config.lora.frequency_offset=0.000
> INFO  | ??:??:?? 5 Set radio: region=EU_868, name=LongFast, config=0, ch=0, power=27
> INFO  | ??:??:?? 5 Radio myRegion->freqStart -> myRegion->freqEnd: 869.400024 -> 869.650024 (0.250000 mhz)
> INFO  | ??:??:?? 5 Radio myRegion->numChannels: 1 x 250.000kHz
> INFO  | ??:??:?? 5 Radio channel_num: 1
> INFO  | ??:??:?? 5 Radio frequency: 869.525024
> INFO  | ??:??:?? 5 Slot time: 42 msec
> INFO  | ??:??:?? 5 Set radio: final power level=22
> INFO  | ??:??:?? 5 SX126x init result 0
> INFO  | ??:??:?? 5 Frequency set to 869.525024
> INFO  | ??:??:?? 5 Bandwidth set to 250.000000
> INFO  | ??:??:?? 5 Power output set to 22
> DEBUG | ??:??:?? 5 Current limit set to 140.000000
> DEBUG | ??:??:?? 5 Current limit set result 0
> DEBUG | ??:??:?? 5 Setting DIO2 as RF switch
> DEBUG | ??:??:?? 5 SX126X_RXEN not defined, defaulting to RADIOLIB_NC
> DEBUG | ??:??:?? 5 SX126X_TXEN not defined, defaulting to RADIOLIB_NC
> DEBUG | ??:??:?? 5 Using MCU pin -1 as RXEN and pin -1 as TXEN to control RF switching
> INFO  | ??:??:?? 5 Set RX gain to power saving mode (boosted mode off); result: 0
> INFO  | ??:??:?? 5 SX1262 Radio init succeeded, using SX1262 radio
> DEBUG | ??:??:?? 5 Initializing MQTT
> DEBUG | ??:??:?? 5 Waiting for SSL Cert to be generated.
> INFO  | ??:??:?? 5 Checking if we have a previously saved SSL Certificate.
> INFO  | ??:??:?? 5 Existing SSL Certificate found!
> DEBUG | ??:??:?? 5 Retrieved Private Key: 1191 Bytes
> DEBUG | ??:??:?? 5 Retrieved Certificate: 792 Bytes
> INFO  | ??:??:?? 5 SSL Cert Ready!
> [  5775][D][WiFiGeneric.cpp:1035] _eventCallback(): Arduino Event: 0 - WIFI_READY
> [  5811][D][WiFiGeneric.cpp:1035] _eventCallback(): Arduino Event: 2 - STA_START
> DEBUG | ??:??:?? 5 JOINING WIFI soon: ssid=
> DEBUG | ??:??:?? 5 (bw=250, sf=11, cr=4/5) packet symLen=8 ms, payloadSize=237, time 2000 ms
> DEBUG | ??:??:?? 5 LoRA bitrate = 118.500000 bytes / sec
> INFO  | ??:??:?? 5 PowerFSM init, USB power=1
> DEBUG | ??:??:?? 5 Enter state: BOOT
> DEBUG | ??:??:?? 5 Setting CPU to 240mhz because WiFi is in use.
> DEBUG | ??:??:?? 5 [Power] Battery: usbPower=0, isCharging=0, batMv=3494, batPct=26
> DEBUG | ??:??:?? 5 [Screen] Screen: Started...
> INFO  | ??:??:?? 5 [RangeTestModule] Range Test Module - Disabled
> INFO  | ??:??:?? 5 [WifiConnect] Reconnecting to WiFi access point 
> INFO  | ??:??:?? 10 [PowerFSM] Loss of power in Powered
> INFO  | ??:??:?? 10 [Screen] Done with boot screen...
> DEBUG | ??:??:?? 10 [Screen] showing standard frames
> DEBUG | ??:??:?? 10 [Screen] Showing 0 module frames
> DEBUG | ??:??:?? 10 [Screen] Total frame count: 103
> DEBUG | ??:??:?? 10 [Screen] Added modules.  numframes: 0
> DEBUG | ??:??:?? 10 [Screen] Finished building frames. numframes: 6
> DEBUG | ??:??:?? 10 [Screen] determineMode(): ghostPixels=0, refresh=FAST, reason=NO_OBJECTIONS, fastRefreshCount=0
> DEBUG | ??:??:?? 10 [Screen] Updating E-Paper... done

**Comment by todd-herbert** [comment]
> Yes, for me it skips the boot screen on both the T-Echo and the Wireless Paper v1.1
That's definitely weird! Maybe it only works on my device because of a different config taking longer to boot?
Thanks a lot for the logs. I'll get onto it today.

**Comment by GUVWAF** [comment]
> That's definitely weird! Maybe it only works on my device because of a different config taking longer to boot?

Seems indeed the case. If I disable Wi-Fi, the boot screen shows on the Wireless Paper.

**Comment by todd-herbert** [comment]
The plan in the near future was to explicitly mark special frames like boot and sleep screens (in Screen.cpp) with `setFrameFlag(COSMETIC)` or `setFrameFlag(DEMAND_FULL)`. 

I think what I'll try is adding a `BLOCKING` flag, and then mark those special frames as `setFrameFlag(COSMETIC & BLOCKING)`, which will tell the EInkDynamicDisplay to sit and repeatedly poll the BUSY pin before returning control to the loop().

**Comment by todd-herbert** [comment]
I'm not able to test it myself, but I believe I do have a fix for the bugs discussed above.
It is tied to code which I had hoped to open a separate pull-request for.

Rather than broadening scope, I am submitting this code as-is, and will quickly draft a new pull-request which handles these "special-case" frames, such as boot and sleep.

**Comment by todd-herbert** [comment]
Please note that this code requires https://github.com/meshtastic/GxEPD2/pull/6 to build correctly

**Comment by GUVWAF** [comment]
> I'm not able to test it myself, but I believe I do have a fix for the bugs discussed above.

Good to hear.

> Rather than broadening scope, I am submitting this code as-is, and will quickly draft a new pull-request which handles these "special-case" frames, such as boot and sleep.

That's fine, it's not a really big issue anyways.

**Review by GUVWAF — APPROVED** [review]

**Review by thebentern — APPROVED** [review]

**Pull request merged**